Corona-Warn-App/src/main/java/de/rki/coronawarnapp/coronatest/TestCertificateRepository.kt

```diff
@@ -0,0 +1,329 @@
+                throw IllegalArgumentException("A certificate was already created for this ${test.identifier}")
+            }
+
+            // TODO do we need additional validation here? last chance to abort?
+
+            val identifier = UUID.randomUUID().toString()
+
+            val certificate = when (test.type) {
+                CoronaTest.Type.PCR -> PCRCertificateContainer(
+                    identifier = identifier,
+                    registeredAt = test.registeredAt,
+                    registrationToken = test.registrationToken,
+                )
+                CoronaTest.Type.RAPID_ANTIGEN -> RACertificateContainer(
+                    identifier = identifier,
+                    registeredAt = test.registeredAt,
+                    registrationToken = test.registrationToken,
+                )
+            }
+            Timber.tag(TAG).d("Adding test certificate entry: %s", certificate)
+            mutate { this[certificate.identifier] = certificate }
+        }
+
+        return newData.values.single { it.registrationToken == test.registrationToken }
+    }
+
+    data class RefreshResult(
+        val certificate: TestCertificateContainer,
+        val error: Exception? = null,
+    )
+
+    suspend fun refresh(identifier: TestCertificateIdentifier? = null): Set<RefreshResult> {
+        Timber.tag(TAG).d("refresh(identifier=%s)", identifier)
+
+        val refreshResults = mutableSetOf<RefreshResult>()
+
+        val workedOnIds = mutableSetOf<TestCertificateIdentifier>()
+
+        internalData.updateBlocking {
+            val toRefresh = values
+                .filter { it.identifier == identifier || identifier == null }
+                .filter { it.isPending }
+
+            mutate {
+                toRefresh.forEach {
+                    workedOnIds.add(it.identifier)
+                    this[it.identifier] = when (it.type) {
+                        CoronaTest.Type.PCR -> (it as PCRCertificateContainer).copy(isUpdatingData = true)
+                        CoronaTest.Type.RAPID_ANTIGEN -> (it as RACertificateContainer).copy(isUpdatingData = true)
+                    }
+                }
+            }
+        }
+
+        internalData.updateBlocking {
+            Timber.tag(TAG).d("Checking for unregistered public keys.")
+
+            val refreshedCerts = values
+                .filter { workedOnIds.contains(it.identifier) }
+                .filter { !it.isPublicKeyRegistered }
+                .map { cert ->
+                    try {
+                        RefreshResult(registerPublicKey(cert))
+                    } catch (e: Exception) {
+                        Timber.tag(TAG).e(e, "Failed to register public key for %s", cert)
+                        RefreshResult(cert, e)
+                    }
+                }
+
+            mutate {
+                refreshedCerts
+                    .filter { it.error == null }
```

Comment:
does this mean that the returned certificates all has errors ?

Reply:
We filter out all `RefreshResult` that had an error so they stay the same as they previously were, while we update the entries that changed.

When refreshing multiple test certificates, each network request could throw an error, but if we successfully registered a public key or even got signed certificate data from the server, then we do not want to throw that away because an operation for a different certificate failed.

I just noticed though that the code is not adding to the top level `refreshResults`, I'll fix that.
The idea is that the top level `refreshResults` contain all operations (with and without error), while we only change the entries in `internalData` for values that were updated without error.